File: sdk/java/src/main/java/com/stratium/sdk/integration/CrossSdkTool.java
package com.stratium.sdk.integration;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.stratium.sdk.client.StratiumClient;
import com.stratium.sdk.client.StratiumClientConfig;
import com.stratium.sdk.client.TokenProvider;
import com.stratium.sdk.client.WrapOptions;
import com.stratium.sdk.client.WrapResult;
import com.stratium.sdk.client.JwtUtils;
import com.stratium.sdk.key.FileKeyStore;

import java.io.IOException;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.util.Base64;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public final class CrossSdkTool {
    private static final ObjectMapper MAPPER = new ObjectMapper()
            .setSerializationInclusion(JsonInclude.Include.NON_NULL);

    private CrossSdkTool() {}

    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.err.println("Usage: CrossSdkTool <encrypt|decrypt>");
            System.exit(1);
        }

        Request request = readRequest();
        try (InteropContext context = new InteropContext()) {
            context.initialize();
            Response response;
            switch (args[0]) {
                case "encrypt" -> response = context.handleEncrypt(request);
                case "decrypt" -> response = context.handleDecrypt(request);
                default -> throw new IllegalArgumentException("Unknown mode: " + args[0]);
            }
            writeResponse(response);
        } catch (Exception e) {
            String message = e.getMessage() != null ? e.getMessage() : e.toString();
            System.err.println(message);
            System.exit(1);
        }
    }

    private static Request readRequest() {
        try {
            return MAPPER.readValue(System.in, Request.class);
        } catch (IOException e) {
            throw new IllegalStateException("Failed to parse JSON request", e);
        }
    }

    private static void writeResponse(Response response) throws IOException {
        MAPPER.writeValue(System.out, response);
    }

    private static final class InteropContext implements AutoCloseable {
        private final Environment env;
        private final StratiumClient client;

        InteropContext() {
            this.env = Environment.fromEnv();
            Path keyDir = resolveKeyDirectory(env.subjectId);
            FileKeyStore keyStore = new FileKeyStore(keyDir);
            this.client = new StratiumClient(env.clientConfig, keyStore, env.tokenProvider);
        }

        void initialize() {
            client.initialize().join();
        }

        Response handleEncrypt(Request req) {
            byte[] plaintext = decodeField(req.plaintext(), "plaintext");
            String policyBase64 = firstNonBlank(req.policyBase64(), env.policyBase64);
            if (policyBase64 == null || policyBase64.isBlank()) {
                throw new IllegalArgumentException("policyBase64 must be provided");
            }

            WrapOptions.Builder builder = WrapOptions.builder()
                    .filename(firstNonBlank(req.filename(), env.defaultFilename))
                    .contentType(firstNonBlank(req.contentType(), env.defaultContentType))
                    .resource(firstNonBlank(req.resource(), env.defaultResource))
                    .policyBase64(policyBase64);

            WrapResult wrapResult = client.wrap(plaintext, builder.build());
            return new Response(Base64.getEncoder().encodeToString(wrapResult.ztdfBlob()), null);
        }

        Response handleDecrypt(Request req) {
            byte[] ztdfBlob = decodeField(req.ztdf(), "ztdf");
            try {
                byte[] plaintext = client.unwrap(ztdfBlob);
                return new Response(null, Base64.getEncoder().encodeToString(plaintext));
            } catch (RuntimeException e) {
                e.printStackTrace(System.err);
                throw e;
            }
        }

        @Override
        public void close() {
            // no-op; StratiumClient manages resources internally
        }
    }

    private static Path resolveKeyDirectory(String subjectId) {
        String override = System.getenv("STRATIUM_JAVA_KEY_DIR");
        if (override != null && !override.isBlank()) {
            return Path.of(override);
        }
        Path cwd = Path.of("").toAbsolutePath();
        Path repoRoot = cwd.getParent() != null ? cwd.getParent().getParent() : cwd;
        if (repoRoot == null) {
            repoRoot = cwd;
        }
        return repoRoot.resolve("artifacts").resolve("client-keys").resolve("java");
    }

    private static final class Environment {
        final StratiumClientConfig clientConfig;
        final TokenProvider tokenProvider;
        final String policyBase64;
        final String defaultResource;
        final String defaultFilename;
        final String defaultContentType;
        final String subjectId;

        private Environment(StratiumClientConfig clientConfig,
                            TokenProvider tokenProvider,
                            String policyBase64,
                            String defaultResource,
                            String defaultFilename,
                            String defaultContentType,
                            String subjectId) {
            this.clientConfig = clientConfig;
            this.tokenProvider = tokenProvider;
            this.policyBase64 = policyBase64;
            this.defaultResource = defaultResource;
            this.defaultFilename = defaultFilename;
            this.defaultContentType = defaultContentType;
            this.subjectId = subjectId;
        }

        static Environment fromEnv() {
            String keyAccessUri = requireEnv("STRATIUM_KEY_ACCESS_URI");
            String keyManagerUri = requireEnv("STRATIUM_KEY_MANAGER_URI");
            String clientId = requireEnv("STRATIUM_CLIENT_ID");
            String realm = requireEnv("STRATIUM_KEYCLOAK_REALM");
            String keycloakUrl = requireEnv("STRATIUM_KEYCLOAK_URL");
            String scope = System.getenv().getOrDefault("STRATIUM_OIDC_SCOPE",
                    "openid profile email stratium-user-profile");
            String clientSecret = System.getenv("STRATIUM_CLIENT_SECRET");
            String username = System.getenv("STRATIUM_USERNAME");
            String password = System.getenv("STRATIUM_PASSWORD");

            TokenProvider tokenProvider = buildTokenProvider(
                    keycloakUrl, realm, clientId, clientSecret, username, password, scope
            );
            String subjectOverride = System.getenv("STRATIUM_SUBJECT_ID");
            String subjectId = (subjectOverride != null && !subjectOverride.isBlank())
                    ? subjectOverride
                    : resolveSubjectId(tokenProvider);

            StratiumClientConfig config = StratiumClientConfig.builder()
                    .keyAccessUri(URI.create(keyAccessUri))
                    .keyManagerUri(URI.create(keyManagerUri))
                    .clientId(clientId)
                    .realm(realm)
                    .subjectId(subjectId)
                    .build();

            String policy = System.getenv("STRATIUM_POLICY_BASE64");
            if (policy == null || policy.isBlank()) {
                throw new IllegalStateException("STRATIUM_POLICY_BASE64 must be set");
            }

            String resource = System.getenv().getOrDefault("STRATIUM_RESOURCE", "integration-resource");
            String filename = System.getenv().getOrDefault("STRATIUM_FILENAME", "interop.txt");
            String contentType = System.getenv().getOrDefault("STRATIUM_CONTENT_TYPE", "text/plain");

            return new Environment(config, tokenProvider, policy, resource, filename, contentType, subjectId);
        }
    }

    private static TokenProvider buildTokenProvider(String keycloakUrl,
                                                    String realm,
                                                    String clientId,
                                                    String clientSecret,
                                                    String username,
                                                    String password,
                                                    String scope) {
        String bearer = System.getenv("STRATIUM_BEARER_TOKEN");
        if (bearer != null && !bearer.isBlank()) {
            final String token = bearer;
            return () -> CompletableFuture.completedFuture(token);
        }

        return com.stratium.sdk.client.KeycloakTokenProvider.builder()
                .authServerUrl(URI.create(keycloakUrl))
                .realm(realm)
                .clientId(clientId)
                .clientSecret(clientSecret)
                .username(username)
                .password(password)
                .scope(scope)
                .build();
    }

    private static String resolveSubjectId(TokenProvider tokenProvider) {
        try {
            String token = tokenProvider.getToken().get(30, TimeUnit.SECONDS);
            return JwtUtils.extractSubject(token);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException("Interrupted while resolving subject id", e);
        } catch (ExecutionException | TimeoutException e) {
            throw new IllegalStateException("Failed to resolve subject id from token", e);
        }
    }

    private static String requireEnv(String key) {
        String value = System.getenv(key);
        if (value == null || value.isBlank()) {
            throw new IllegalStateException("Environment variable " + key + " is required");
        }
        return value;
    }

    private static byte[] decodeField(String value, String name) {
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException("Missing " + name + " field");
        }
        try {
            return Base64.getDecoder().decode(value);
        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException("Invalid base64 in " + name, e);
        }
    }

    private static String firstNonBlank(String primary, String fallback) {
        if (primary != null && !primary.isBlank()) {
            return primary;
        }
        return fallback;
    }

    private record Request(String plaintext,
                           String ztdf,
                           String filename,
                           String contentType,
                           String resource,
                           String policyBase64) {
    }

    private record Response(String ztdf, String plaintext) {
    }
}

File: sdk/js/scripts/cross-sdk-tool.mjs
#!/usr/bin/env node

import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';
import { ZtdfClient } from '../src/nodejs/ztdf-client.js';
import { setStorageDirectory } from '../src/nodejs/key-storage.js';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

async function main() {
  const mode = process.argv[2];
  if (!mode || !['encrypt', 'decrypt'].includes(mode)) {
    console.error('Usage: node cross-sdk-tool.mjs <encrypt|decrypt>');
    process.exit(1);
  }

  const rawInput = await readStdin();
  if (!rawInput) {
    throw new Error('Missing JSON payload on stdin');
  }

  let request;
  try {
    request = JSON.parse(rawInput);
  } catch (err) {
    throw new Error(`Failed to parse JSON request: ${err.message}`);
  }

  const env = loadEnv();
  await ensureDir(env.keyDir);
  const client = new ZtdfClient({
    keyAccessUrl: env.keyAccessUrl,
    keyManagerUrl: env.keyManagerUrl,
    clientId: env.clientId,
    getToken: () => fetchToken(env),
    debug: env.debug,
  });

  setStorageDirectory(env.keyDir);
  await client.initialize();

  if (mode === 'encrypt') {
    const resp = await handleEncrypt(client, env, request);
    writeResponse(resp);
  } else {
    const resp = await handleDecrypt(client, env, request);
    writeResponse(resp);
  }
}

async function handleEncrypt(client, env, request) {
  const plaintext = decodeField(request.plaintext, 'plaintext');
  const options = {
    filename: request.filename || env.defaultFilename,
    contentType: request.contentType || env.defaultContentType,
    resource: request.resource || env.defaultResource,
    policy: request.policyBase64 || env.policyBase64,
  };
  if (!options.policy) {
    throw new Error('policyBase64 must be provided');
  }

  const ztdfBuffer = await client.wrap(Buffer.from(plaintext), options);
  return { ztdf: ztdfBuffer.toString('base64') };
}

async function handleDecrypt(client, env, request) {
  const ztdfBytes = decodeField(request.ztdf, 'ztdf');
  const result = await client.unwrap(Buffer.from(ztdfBytes), {
    resource: request.resource || env.defaultResource,
  });
  return { plaintext: Buffer.from(result.content).toString('base64') };
}

function decodeField(value, fieldName) {
  if (!value || typeof value !== 'string') {
    throw new Error(`Missing ${fieldName} field`);
  }
  try {
    return Buffer.from(value, 'base64');
  } catch (err) {
    throw new Error(`Invalid base64 in ${fieldName}: ${err.message}`);
  }
}

function writeResponse(obj) {
  process.stdout.write(`${JSON.stringify(obj)}\n`);
}

function readStdin() {
  return new Promise((resolve, reject) => {
    let data = '';
    process.stdin.setEncoding('utf8');
    process.stdin.on('data', (chunk) => {
      data += chunk;
    });
    process.stdin.on('end', () => resolve(data.trim()));
    process.stdin.on('error', reject);
  });
}

function loadEnv() {
  const required = (key) => {
    const value = process.env[key];
    if (!value) {
      throw new Error(`Environment variable ${key} is required`);
    }
    return value;
  };

  const repoRoot = path.resolve(__dirname, '..', '..', '..');
  const keyDir = process.env.STRATIUM_JS_KEY_DIR
    || path.join(repoRoot, 'artifacts', 'client-keys', 'js');

  const keyAccessUrl = resolveUrl(
    process.env.STRATIUM_KEY_ACCESS_URI,
    process.env.STRATIUM_KEY_ACCESS_URL,
    process.env.STRATIUM_KEY_ACCESS_ADDR
  );
  const keyManagerUrl = resolveUrl(
    process.env.STRATIUM_KEY_MANAGER_URI,
    process.env.STRATIUM_KEY_MANAGER_URL,
    process.env.STRATIUM_KEY_MANAGER_ADDR
  );

  return {
    keyAccessUrl,
    keyManagerUrl,
    clientId: required('STRATIUM_CLIENT_ID'),
    keycloakUrl: required('STRATIUM_KEYCLOAK_URL'),
    realm: required('STRATIUM_KEYCLOAK_REALM'),
    clientSecret: process.env.STRATIUM_CLIENT_SECRET || '',
    username: process.env.STRATIUM_USERNAME || '',
    password: process.env.STRATIUM_PASSWORD || '',
    bearerToken: process.env.STRATIUM_BEARER_TOKEN || '',
    policyBase64: process.env.STRATIUM_POLICY_BASE64 || '',
    defaultResource: process.env.STRATIUM_RESOURCE || 'integration-resource',
    defaultFilename: process.env.STRATIUM_FILENAME || 'interop.txt',
    defaultContentType: process.env.STRATIUM_CONTENT_TYPE || 'text/plain',
    keyDir,
    debug: process.env.STRATIUM_SDK_DEBUG === '1',
  };
}

async function ensureDir(dir) {
  try {
    await fs.mkdir(dir, { recursive: true, mode: 0o700 });
  } catch (err) {
    if (err.code !== 'EEXIST') {
      throw err;
    }
  }
}

function resolveUrl(...candidates) {
  for (const value of candidates) {
    if (!value) continue;
    if (value.startsWith('http://') || value.startsWith('https://')) {
      return value;
    }
    // No scheme, try to prepend http://
    if (/^[-\w.]+:\d+$/.test(value) || value.startsWith('localhost')) {
      return `http://${value}`;
    }
    return value;
  }
  throw new Error('Missing required URL/URI value');
}

async function fetchToken(env) {
  if (env.bearerToken) {
    return env.bearerToken;
  }

  const tokenUrl = `${env.keycloakUrl}/realms/${env.realm}/protocol/openid-connect/token`;
  const params = new URLSearchParams();
  params.set('client_id', env.clientId);
  if (env.clientSecret) {
    params.set('client_secret', env.clientSecret);
  }

  if (env.username && env.password) {
    params.set('grant_type', 'password');
    params.set('username', env.username);
    params.set('password', env.password);
  } else {
    params.set('grant_type', 'client_credentials');
  }

  const res = await fetch(tokenUrl, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: params.toString(),
  });
  if (!res.ok) {
    throw new Error(`Token request failed: ${res.status} ${res.statusText}`);
  }
  const data = await res.json();
  if (!data.access_token) {
    throw new Error('Token response missing access_token');
  }
  return data.access_token;
}

main().catch((err) => {
  const message = err?.message || String(err);
  console.error(message);
  process.exit(1);
});
